pkg/ingester/ingester_v2_backfill.go

```diff
@@ -0,0 +1,499 @@
+
+	// Keeping this methods of Ingester here to avoid the appender having a reference to the ingester.
+	createNewTSDB func(userID, dbDir string, minBlockDuration, maxBlockDuration int64, reg *prometheus.Registry) (*userTSDB, error)
+	// We are keeping the method instead of string for backfill dir because if there is a surge in backfill writes,
+	// it will create a new string for every push while we want it only to open a new TSDB.
+	backfillBlocksDir func(userID string) string
+}
+
+type backfillRangeAppender struct {
+	app        storage.Appender
+	db         *userTSDB
+	start, end int64
+}
+
+func newBackfillRangeAppender(db *backfillTSDBWrapper) *backfillRangeAppender {
+	return &backfillRangeAppender{
+		app:   db.db.Appender(),
+		db:    db.db,
+		start: db.start,
+		end:   db.end,
+	}
+}
+
+// add requires the samples to be within the backfill age.
+// Note: This method should not be called concurrently.
+func (a *backfillAppender) add(la []client.LabelAdapter, s client.Sample) (err error) {
+	app, db, err := a.getAppender(s)
+	if err != nil {
+		return err
+	}
+	startAppend := time.Now()
+
+	cachedRef, cachedRefExists := db.refCache.Ref(startAppend, client.FromLabelAdaptersToLabels(la))
+	// If the cached reference exists, we try to use it.
+	if cachedRefExists {
+		err = app.AddFast(cachedRef, s.TimestampMs, s.Value)
+		if err == nil || errors.Cause(err) != storage.ErrNotFound {
+			return err
+		}
+	}
+
+	// Copy the label set because both TSDB and the cache may retain it.
+	copiedLabels := client.FromLabelAdaptersToLabelsWithCopy(la)
+	ref, err := app.Add(copiedLabels, s.TimestampMs, s.Value)
+	if err == nil {
+		db.refCache.SetRef(startAppend, copiedLabels, ref)
+	}
+
+	return err
+}
+
+func (a *backfillAppender) getAppender(s client.Sample) (storage.Appender, *userTSDB, error) {
+	if !a.backfillTSDB.isWithinBackfillAge(s.TimestampMs) {
+		return nil, nil, storage.ErrOutOfBounds
+	}
+
+	// Fast path to not take the lock.
+	if a.newerAppender != nil && s.TimestampMs >= a.newerAppender.start && s.TimestampMs < a.newerAppender.end {
+		return a.newerAppender.app, a.newerAppender.db, nil
+	} else if a.olderAppender != nil && s.TimestampMs >= a.olderAppender.start && s.TimestampMs < a.olderAppender.end {
+		return a.olderAppender.app, a.olderAppender.db, nil
+	}
+
+	// We could use RLock here to check the TSDBs, but because we create new TSDB if none exists,
+	// it's easier and cleaner to handle races by taking the write lock.
+	// TODO(codesome): If we see performance issues, explore ways to read lock here.
+	// TODO(codesome): IMPORTANT: fix the issue where the compaction moves the TSDB around between 2 appends to avoid using wrong appenders in older and newer spots.
+	a.backfillTSDB.mtx.Lock()
+	defer a.backfillTSDB.mtx.Unlock()
+
+	// Check if we already have TSDB created and use it.
+	if a.backfillTSDB.dbs[0] != nil && s.TimestampMs >= a.backfillTSDB.dbs[0].start && s.TimestampMs < a.backfillTSDB.dbs[0].end {
```

Comment:
You could move this long line logic into a simple function on `backfillTSDBWrapper.within(s.TimestampMs)`